docs/client-conventions.md

```diff
@@ -0,0 +1,164 @@
+# Client Conventions
+
+This document describes conventions that Knative domain-specific clients can
+follow to achieve specific end-user goals. It is intended as a set of best
+practices for client implementors, and also as advice to direct users of the API
+(for example, with kubectl).
+
+These conventions are merely conventions:
+ * They are optional; you can use Knative entirely validly without them.
+ * They are designed to be useful even when some clients are not obeying the
+   conventions. Each convention describes what happens in the presence of
+   convention-unaware clients.
+
+Some of the conventions involve the client setting labels or annotations; the
+`client.knative.dev/*` label/annotation namespace is reserved for documented
+Knative client convetions.
+
+## Determine when an action is complete
+
+As Knative is (like all of Kubernetes) a declarative API, the user expresses
+their desire by changing some values in the Knative objects. Clients need not be
+declarative, and might have expressions of user intent like "Deploy this code"
+or "Change these environment variables". To tell when such an action is
+complete, the client can look at the status conditions.
+
+Each Knative object has a `Ready` status condition. When a change is initiated,
+the controller flips this to `Unknown`. When the serving state again reflects
+exactly what the spec of the object specifies, the `Ready` condition will flip
+to `True`; this indicates the operation was a success. If reflecting the spec in
+the serving state is impossible, the `Ready` condition will flip to `False`;
+this indicates the operation was a failure, and the message of the status
+condition should indicate something in English about why (and the Reason field
+can indicate an emumeration suitable for i18n). Either `True` or `False`
+indicates the operation is complete, for better or worse.
+
+Note that someone else could start another operation while the cient was waiting
+for its operation. A conventional client still waits for the `Ready` condition
+to land at `True` or `False`, and then describes to the user what happened using
+logic based on the intended effect.
+
+For example:
+ * Client A deploys image `gcr.io/foods/vegetables:eggplant`
+ * While that is not yet Ready, client B deploys `gcr.io/foods/vegetables:squash`
+ * The `eggplant` revision becomes Ready: True, and the service moves traffic to it.
```

Comment:
Technically `knative/serving` doesn't do this today[1], but I don't see any reason that a conforming implementation couldn't.

[1] - We only update `latestReadyRevision = latestCreatedRevision` today.